lib/charms/nrpe_exporter/v0/nrpe_exporter.py

```diff
@@ -160,6 +168,19 @@ def _validate_relation_by_interface_and_direction(
         raise Exception("Unexpected RelationDirection: {}".format(expected_relation_role))
 
 
+def _type_convert_stored(obj):
+    """Convert Stored* to their appropriate types, recursively."""
+    if isinstance(obj, StoredList):
+        return list(map(_type_convert_stored, obj))
+    elif isinstance(obj, StoredDict):
+        rdict = {}  # type: Dict[Any, Any]
+        for k in obj.keys():
+            rdict[k] = _type_convert_stored(obj[k])
+        return rdict
+    else:
+        return obj
```

Comment:
I mean, technically speaking, it does, thanks to fun/no-fun pass-by-reference semantics which are hidden from users.

In practice, though, this function is used all over the place. It came out of the OF "make `Stored* types behave like the objects they claim to be" PR which got rejected, as both an example of how simple it is to cast them back, and how impossible it is to do anything otherwise.

When these come back from the Juju controller (where they are inevitably JSON-serialized simple types which would not have been infinitely recursive), something like:

```python
{
  "foo": [
      {
          "bar": ["quux"]
      }
   ]
}
```

Comes back out as:

`StoredDict[str, StoredList[StoredDict[str, StoredList[str]`

And a "naive" cast (`dict(SomeStoredDict)`) doesn't recurse to cast the nested objects, so trying to JSON/YAML serialize them doesn't behave like authors would expect.

The point of the function is basically unwrapping this, but since we're basically assured that it's coming from Juju as `map[string]string` (instead of `map[str]interface{}`) because it's more or less a simple k:v store on the backend, the odds of a charm author wrapping an infinitely recursive Python object are almost zero, and would have to be intentional. The `isinstance()` checks in the example, anyway (which would just be plain `dict|list` and not `StoredDict|StoredList`, which are only instantiated when `StoredData` is pulled from the controller) wouldn't recurse through your example at all, and would just return the original.

I guess the converse case is an author doing something like:

```python
{
  "data": self._stored.somedict
}
```

And passing it to that method, which would not in fact convert the StoredDict under a "normal" key, but that's been an ok tradeoff so far.